### PR TITLE
Mods for energy diagnostics

### PR DESCRIPTION
**Pull request by oksanaguba**
Adding (disabled by default) diagnostics for the pressure work (moist adjustment) term.
Reorganizing energy code in eam, to avoid repetition.
Removing unnecessary allocates.
Adding infrastructure to use homme/src/theta-l for debugging.

[bfb]

### CODE REVIEW

**Comment by rljacob** [comment]
@oksanaguba can you please clean up this branch and combine commits that just say "cleanup"?

**Comment by oksanaguba** [comment]
@rljacob how would you want me to do that? squash them? To clarify, there is a reason for many clean up commits -- I ran a bfb  test between each of them. To save my work in between, I used git commit instead of some other mechanism..

**Comment by rljacob** [comment]
Yes squash them.  That's a perfectly fine dev practice but when it comes time to merge to master, we only want substantive commit messages in the history.

**Comment by AaronDonahue** [comment]
@oksanaguba , I'm just now checking back on this PR.  Have you had a chance to squash the commit history and write up a description of the PR?

**Comment by rljacob** [comment]
@oksanaguba please fix this PR.  Squash commits, write a description.

**Comment by oksanaguba** [comment]
@rljacob i tried to squash but i keep getting an error. there is a commit, master merge into this branch, that rebase cannot apply.

**Comment by oksanaguba** [comment]
@AaronDonahue sorry i missed your message, i will try to finalize this PR.

**Comment by oksanaguba** [comment]
@mt5555 if you could please review changes in prim_state for energy output?

**Comment by rljacob** [comment]
@mt5555 please review this.

**Comment by mt5555** [comment]
This code is adding advanced energy budget diagnostics to the physics - I'm not familiar enough with that code to be a reviewer. Suggest @singhbalwinder

**Comment by oksanaguba** [comment]
@mt5555 could you please look at homme files only?

File: components/homme/src/share/control_mod.F90
Comment (mt5555) [inline components/homme/src/share/control_mod.F90:616]: cound this code be moved into the model specific model_init_mod.F90?  that would let us get rid of the #ifdef MODEL_THETA_L stuff.  (I'm always trying to reduce #ifdef's )

**Review by mt5555 — APPROVED** [review]

**Comment by AaronDonahue** [comment]
@oksanaguba and @mt5555 , are we okay to start merging this without the review from @singhbalwinder?

**Comment by oksanaguba** [comment]
@AaronDonahue No, please, let him look. I want to do another check in this PR but I was waiting to see what the reviewers say.

**Comment by AaronDonahue** [comment]
@oksanaguba , great, I think he was pinged with my comment so I'll wait for his review to come in.  Thanks for responding.

**Comment by singhbalwinder** [comment]
I am so sorry about the delay. I started looking at it and will finish the review soon.

File: components/cam/src/physics/cam/check_energy.F90
Comment (singhbalwinder) [inline components/cam/src/physics/cam/check_energy.F90:357]: I may be wrong but from my reading of the code, it looks like cpairv_loc can be different from cpair for WACCM simulations. I haven't run the code in WACM mode to verify that.

File: components/cam/src/physics/cam/check_energy.F90
Comment (singhbalwinder) [inline components/cam/src/physics/cam/check_energy.F90:452]: Thanks for encapsulating this repeated code in a subroutine!

File: components/cam/src/physics/cam/physics_types.F90
Comment (singhbalwinder) [inline components/cam/src/physics/cam/physics_types.F90:238]: I also like removing runtime allocations-deallocations if not absolutely needed. These are expensive calls and sometime defragments the memory.

File: components/cam/src/physics/cam/physics_types.F90
Comment (singhbalwinder) [inline components/cam/src/physics/cam/physics_types.F90:395]: Are we getting rid of WACCM related codes? I am not familiar with WACCM, so I am not sue how WACCM simulations will be impacted by this change.

**Review by singhbalwinder — CHANGES_REQUESTED** [review]
Thanks @oksanaguba for adding this feature. I left some comments based on my understanding of the code. Please feel free to push back if you think my suggested changes may cause other issues.

**Comment by rljacob** [comment]
@oksanaguba please respond to @singhbalwinder 's review comments.

**Comment by oksanaguba** [comment]
I cannot work on this for another few weeks, closing for now.

**Comment by oksanaguba** [comment]
@wlin7 -- I would like to ask you about this PR. I want to revive a part of it which removes cpair arrays (not used since we have const pair for now) and repeated energy calculations, so that energy definition is in one routine only. cpairv, rairv (zvirv) arrays are used in WACCM, but is anyone actually using this in e3sm? Besides asking you, I will ask Yaga R. and Jack in case they use WACCM options.

**Pull request closed without merging**
